### DISCUSSION TITLE
Why does the path "/posts/[slug]" get pre-rendered? (or, what is a suggested pre-13 strategy for loading site-wide data?)

### DISCUSSION THREAD

**Discussion by good-idea** [vercel/next.js#46074] [Help]
I have a next.js v12 build that requires all pages fetch data needed for site-wide things, such as a header. To do this, all of my `getStaticProps` look something like this:

```ts
export const getStaticProps: GetStaticProps<PostProps> = async (ctx) => {
  const { params } = ctx
  if (!params?.slug || typeof params.slug !== 'string') {
    return { notFound: true }
  }

  const [postData, siteSettings] = await Promise.all([
    client.getPost(params.slug),
    fetchSiteSettings(), 👈 global data
  ])
  if (!postData) {
    return {
      notFound: true,
      props: { siteSettings },
    }
  }
  return { props: { postData, siteSettings }, revalidate: 10 }
}
```

In `_app.tsx`, I assert that it has been fetched:

```ts
export default function App({ Component, pageProps, router }: AppProps) {
  const { siteSettings } = pageProps
  if (!siteSettings) {
    throw new Error(
      `[route: ${router.asPath}] Site settings were not fetched. Did you forget to include fetchSiteSettings in getStaticProps?`,
    )
  }
  return (...)
}
```

However, when I run `yarn build`, it throws when attempting to generate the **path** `posts/[slug]`: `Error: [route: /posts/[slug]] Site settings were not fetched. Did you forget to include fetchSiteSettings in getStaticProps?`

That page's `getStaticPaths` uses `{ fallback: true }`; I can see that setting this to `'blocking'` fixes the issue.

[Here is a reproduction](https://github.com/good-idea/next-ssr-param-error-example).

My questions:

- With `fallback: true` - why is the path `posts/[slug]` being pre-rendered in the first place?
  - Is there a way to fetch this `siteSettings` data when that is happening?
- Other than the load time that a user would encounter, is there any reason to not use `{ fallback: 'blocking' }`?
- Any suggestions on other ways to fetch global data other than leveraging the new `app` directory?

**Comment by icyJoseph** [comment]
Global data fetching, as in for all pages, is very tricky... If the user arrives to your page from anything else than the page(s) that injects the global data, then your UI will look funky.

The way to solve this, if you must do SSR for the header and other elements that use this global data, is to use _app `getInitialProps`, and detect server calls to get the data you need. On client side calls you can:

- skip passing props, 
- use context, which skips updating if the incoming props are undefined/null
- or read from __NEXT_DATA__ to re-inject the same data back into the page, this is a bit slower, because you ought to serialize JSON data.

This would disable pre-rendering for pages that do not use `getStaticProps` though.

Onto your questions.

> With fallback: true - why is the path posts/[slug] being pre-rendered in the first place?
> Is there a way to fetch this siteSettings data when that is happening?

Easy, when fallback is `true` and the user navigates onto a page that's not yet generated, we can either show a blank, (`"blocking"`), or show a loading/fallback UI (`true`), so Next.js pre-renders the loading/fallback UI during build time.

When pre-rendering the loading UI data props are undefined, and you have to check the `useRouter().isFallback` property to guide the SSR pre-rendering.

This is all well documented, with code examples and the same reasoning, here: https://nextjs.org/docs/api-reference/data-fetching/get-static-paths#fallback-pages
 
> Other than the load time that a user would encounter, is there any reason to not use { fallback: 'blocking' }?

Well, the user gets to see a snappier UI, because the loading/fallback UI kicks in quickly. There's two situations, you are already on the APP and navigate to the page with GSP + fallback, or you arrive from another site to GSP + fallback

In the first case, on transition, next/link fetches both the JavaScript and data needed by asking for GSP server side, the response is used as props for the JavaScript that runs the page. Even blocking would benefit from this, unless the GSP call takes very long... but pre-fetching is usually done when the next/link is within the viewport.

In the second case, blocking has to wait for:

- gSP to run server side
- HTML to be generated

While fallback true, just serves the fallback UI immediately.

The fallback UI is not served to bots.

> Any suggestions on other ways to fetch global data other than leveraging the new app directory?

As said, _app `getInitialProps`, or wrap every page with a data fetching Higher order function (hard to maintain).
